mathjax3-ts/components/latest.ts

```diff
@@ -0,0 +1,398 @@
+
+/**
+ * Request JSON data from a CDN.  If it loads OK, call the action() function
+ * on the data.  If not, or if the action returns false, run the failure() function.
+ *
+ * @param {CdnData} cdn        The CDN whose API will be used
+ * @param {Function} action    The function to perfrom when the data are received
+ * @param {Function} failure   The function to perform is that data can't be obtained,
+ *                               or if action() returns false
+ */
+function requestXML(cdn: CdnData, action: (json: JSON | JSON[]) => boolean, failure: () => void) {
+    var request = getXMLHttpRequest();
+    if (request) {
+        request.onreadystatechange = function () {
+            if (request.readyState === 4) {
+                if (request.status === 200) {
+                    !action(JSON.parse(request.responseText)) && failure();
+                } else {
+                    Error('Problem acquiring MathJax version: status = ' + request.status);
+                    failure();
+                }
+            }
+        };
+        request.open('GET', cdn.api, true);
+        request.send(null);
+    } else {
+        Error('Can\'t create XMLHttpRequest object');
+        failure();
+    }
+}
+
+/**
+ * Look through the list of versions on GitHub and find the first one that
+ * has the MJX_VERSION as its major version number, and load that.  If none
+ * is found, run the version from which latest.js was loaded.
+ */
+function loadLatestGitVersion() {
+    requestXML(GITHUB, (json: JSON[]) => {
+        if (!(json instanceof Array)) return;
+        for (const data of json) {
+            if (checkVersion((data as any)[GITHUB.key])) {
+                return true;
+            }
+        }
+        return false;
+    }, loadDefaultMathJax);
+}
+
+/**
+ * Check the CDN for its latest version, and load that, if it is an
+ * acceptable version, otherwise, (e.g., the current version has a
+ * higher major version that MJX_VERSION), find the highest version on
+ * GitHub with the given major version and use that.  If one can't be
+ * found, use the version where latest.js was loaded.
+ */
+function loadLatestCdnVersion() {
+    requestXML(script.cdn, function (json) {
+        if (json instanceof Array) {
+            json = json[0];
+        }
+        if (!checkVersion((json as any)[script.cdn.key])) {
+            loadLatestGitVersion();
+        }
+        return true;
+    }, loadDefaultMathJax);
+}
+
+/*=====================================================================*/
+
+
+/*
+ * Find the script that loaded latest.js
+ * If the script is from a known CDN:
+ *   Retrieve the cached version (if any)
+ *   Load the given version of the file, if the version is cached,
+ *   Otherwise find the latest version and load that.
+ * Otherwise,
+ *   Load using the version where latest.js was loaded.
+ */
+export function loadLatest() {
+    script = getScript();
+    if (script && script.cdn) {
+        var version = getSavedVersion();
```

Comment:
```suggestion
        const version = getSavedVersion();
```

Reply:
Right.  I originally had this file as a plain javascript one (where I didn't want to use ES6 for backward compatibility), then decided to convert to typescript.  Missed changing that one.